lisp/bazel-build.el

```diff
@@ -28,35 +28,111 @@
 
 (defun bazel-build (target)
   "Build a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel build ")))
+  (interactive (list (bazel-build--read-target "bazel build")))
   (bazel-build--run-bazel-command "build" target))
 
 (defun bazel-run (target)
   "Build and run a Bazel TARGET."
-  (interactive (list (bazel-build--read-target "bazel run ")))
+  (interactive (list (bazel-build--read-target "bazel run")))
   (bazel-build--run-bazel-command "run" target))
 
 (defun bazel-test (target)
   "Build and run a Bazel test TARGET."
-  (interactive (list (bazel-build--read-target "bazel test ")))
+  (interactive (list (bazel-build--read-target "bazel test")))
   (bazel-build--run-bazel-command "test" target))
 
+(defvar bazel-command "bazel"
+  "Bazel executable to run for subprocesses.")
+
+(defvar bazel-query-args '("query" "--noblock_for_lock")
+  "Bazel query subcommand and arguments.")
+
+(defvar bazel-run-commands-from-root nil
+  "A flag indicating to run commands from the workspace root.")
+
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
-  (compile
-   (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
-
-(defun bazel-build--read-target (prompt)
-  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
-  (let* ((file-name (buffer-file-name))
-         (workspace-root
-          (or (bazel-util-workspace-root file-name)
-              (user-error "Not in a Bazel workspace.  No WORKSPACE file found")))
-         (package-name
-          (or (bazel-util-package-name file-name workspace-root)
-              (user-error "Not in a Bazel package.  No BUILD file found")))
-         (initial-input (concat "//" package-name)))
-    (read-string prompt initial-input)))
+  (let ((command
+         (if bazel-run-commands-from-root
+             (let ((root-dir (or (bazel-util-workspace-root
+                                  (or (buffer-file-name)
+                                      (and dired-directory (directory-file-name dired-directory))))
+                                 (user-error "Could not find workspace."))))
+               (format "cd %s && %s %s %s" root-dir
+                       bazel-command command target))
+           (mapconcat #'shell-quote-argument (list bazel-command command target) " "))))
+    (compile command)))
+
+(defun bazel-build--call-process (&rest args)
+  "Run a bazel subcommand. Return stdout as string."
+  ;; Note: The current working directory of the subprocess is set to the current
+  ;; buffer's value of default-directory.
+  (let ((stderr-file (make-temp-file "bazel-stderr")))
+    (unwind-protect
+        (string-rstrip
+         (with-output-to-string
+           (let ((status
+                  (apply #'call-process bazel-command
```

Comment:
Could you make this work remotely, by using process-file?

Reply:
Done. 
I'm not super familiar with other differences, but I read the docstring, the interface seems to be the same as well, and I used it and it seems to work just the same.